release-23.1: server/license: Add core throttling logic for license policy enforcement

spilchen opened the pull request:
Backport 1/1 commits from #129880.

/cc @cockroachdb/release

---

As part of deprecating the CockroachDB core license, new logic has been added to throttle access when license policies are violated. This change introduces the core logic, though it does not yet integrate with the SQL engine, so throttling won't affect SQL users at this stage.

The LicenseEnforcer has been enhanced with throttling logic, including a new API, MaybeFailIfThrottled, which determines whether throttling should occur and returns an error if it is active.

To make throttling decisions, the LicenseEnforcer requires access to license state. However, since the license code remains in the CCL package, which cannot be imported directly into the server code, a workaround is used. When the enterprise.license changes, a callback (from CCL to server packages) updates the state. Initially, this callback points to a stub function, which the CCL package replaces with the actual implementation upon import.

The implementation also includes handling for mandatory telemetry data required by certain licenses. While the logic is in place to make decisions based on the last telemetry ping, the integration with the diagnostics reporter for actual telemetry ping times will be completed in a follow-up change.

This change will be backported to 23.1, 23.2, 24.1 and 24.2.

Epic: CRDB-39988
Informs: CRDB-39991
Release note: None
Release justification: This work is part of the CockroachDB core deprecation.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/130352)
<!-- Reviewable:end -->

spilchen committed:
server/license: Add core throttling logic for license policy enforcement

As part of deprecating the CockroachDB core license, new logic has been
added to throttle access when license policies are violated. This change
introduces the core logic, though it does not yet integrate with the SQL
engine, so throttling won't affect SQL users at this stage.

The LicenseEnforcer has been enhanced with throttling logic, including a
new API, MaybeFailIfThrottled, which determines whether throttling
should occur and returns an error if it is active.

To make throttling decisions, the LicenseEnforcer requires access to
license state. However, since the license code remains in the CCL
package, which cannot be imported directly into the server code, a
workaround is used. When the enterprise.license changes, a callback
(from CCL to server packages) updates the state. Initially, this
callback points to a stub function, which the CCL package replaces with
the actual implementation upon import.

The implementation also includes handling for mandatory telemetry data
required by certain licenses. While the logic is in place to make
decisions based on the last telemetry ping, the integration with the
diagnostics reporter for actual telemetry ping times will be completed
in a follow-up change.

This change will be backported to 23.1, 23.2, 24.1 and 24.2.

Epic: CRDB-39988
Informs: CRDB-39991
Release note: None
